src/Tasks/Unzip.cs

```diff
@@ -0,0 +1,193 @@
+                                return false;
+                            }
+                        }
+                    }
+                }
+                catch (OperationCanceledException)
+                {
+                    break;
+                }
+                catch (Exception e)
+                {
+                    // Should only be thrown if the archive could not be opened (Access denied, corrupt file, etc)
+                    Log.LogErrorFromResources("Unzip.ErrorCouldNotOpenFile", sourceFile.ItemSpec, e.Message);
+                }
+            }
+
+            BuildEngine3.Reacquire();
+
+            return !_cancellationToken.IsCancellationRequested && !Log.HasLoggedErrors;
+        }
+
+        /// <summary>
+        /// Extracts all files to the specified directory.
+        /// </summary>
+        /// <param name="sourceArchive">The <see cref="ZipArchive"/> containing the files to extract.</param>
+        /// <param name="destinationDirectory">The <see cref="DirectoryInfo"/> to extract files to.</param>
+        private void Extract(ZipArchive sourceArchive, DirectoryInfo destinationDirectory)
+        {
+            foreach (ZipArchiveEntry zipArchiveEntry in sourceArchive.Entries.TakeWhile(i => !_cancellationToken.IsCancellationRequested))
+            {
+                FileInfo destinationPath = new FileInfo(Path.Combine(destinationDirectory.FullName, zipArchiveEntry.FullName));
+
+                if (!destinationPath.FullName.StartsWith(destinationDirectory.FullName, StringComparison.OrdinalIgnoreCase))
+                {
+                    // ExtractToDirectory() throws an IOException for this but since we're extracting one file at a time
+                    // for logging and cancellation, we need to check for it ourselves.
+                    Log.LogErrorFromResources("Unzip.ErrorExtractingResultsInFilesOutsideDestination", destinationPath.FullName, destinationDirectory.FullName);
+                    continue;
+                }
+
+                if (ShouldSkipEntry(zipArchiveEntry, destinationPath))
+                {
+                    Log.LogMessageFromResources(MessageImportance.Low, "Unzip.DidNotUnzipBecauseOfFileMatch", zipArchiveEntry.FullName, destinationPath.FullName, nameof(SkipUnchangedFiles), "true");
+                    continue;
+                }
+
+                try
+                {
+                    destinationPath.Directory?.Create();
+                }
+                catch (Exception e)
+                {
+                    Log.LogErrorWithCodeFromResources("Unzip.ErrorCouldNotCreateDestinationDirectory", destinationPath.DirectoryName, e.Message);
+                    continue;
+                }
+
+                if (OverwriteReadOnlyFiles && destinationPath.Exists && destinationPath.IsReadOnly)
+                {
+                    try
+                    {
+                        destinationPath.IsReadOnly = false;
```

Comment:
Should it be made readonly after the write to preserve user data? :)

Reply:
I'm not sure.  I'd assume that if the file is readonly in the archive it should be readonly after extraction but I don't see attributes like that in the `ZipArchiveEntry` class...  It seems a little weird to be setting the readonly attribute based on what the file _used_ to have.  Thoughts?